### PR TITLE
Add Docker Compose development environment

### LINKED ISSUE #287: Add docker compose local development environment
For faster dev spinup, let's add a container based development environment (this was requested by contributors).

### PR DESCRIPTION
**Pull request by redshiftzero**
## Status

Ready for review

## Description of Changes

Fixes #287

### CODE REVIEW

**Comment by redshiftzero** [comment]
Test failures are unrelated and are fixed in #338 (that one will be merged first)

**Comment by r4v5** [comment]
I've tested this, and it's cool in theory but i'm getting
> web_1       | OSError: [Errno 13] Permission denied: '/usr/src/app/OpenOversight'

(the shared-in volume doesn't have the right permissions for www-data to read it) on my linux machine. I know this is probably a "me" problem more than a "everyone" problem, but it makes running this on linux a lot harder than it should be.

**Comment by r4v5** [comment]
also: gunicorn 19.0 added a `--reload` option! it live-reloads your code so you can see changes in dev without killing and restarting/rebuilding your server! I assert this is worth the upgrade.

**Comment by redshiftzero** [comment]
Thanks for giving this a whirl @r4v5 - will take another look since we def want this to work well on Linux :-)

**Comment by pierwill** [comment]
This is **sweet**! 

In a debian-9 vm, I followed [these instructions](https://docs.docker.com/install/linux/docker-ce/debian/) for docker-ce, then did `sudo apt-get install docker-compose` to get version 1.8. (Did I need to do both?) However, I had to change `version: "3"` to `version: "2"` in ` docker-compose.yml` to get things working.

**Comment by b-meson** [comment]
Approved for OS X.  @elimisteve will test on Qubes / debian 9

**Comment by tmc** [comment]
can we move the postgres port to be other than 5432? as is I can't also be running postgres.

**Comment by redshiftzero** [comment]
I'm cool with moving that port

**Comment by elimisteve** [comment]
I agree that changing the OO Postgres port from 5432 would be nice

**Comment by b-meson** [comment]
There is a corresponding documentation change that needs to be updated in the file `DOCKER_DEPLOY.md`. I honestly think we can just drop that file and put it all in `CONTRIB.md`

**Comment by elimisteve** [comment]
```
ERROR: for web  Cannot start service web: driver failed programming external connectivity on endpoint openoversight_web_1 (f550dbb0478bb473057c607ac8eb55d5f3d6f2b43e719b36179814bca7c7b09a): Error starting userland proxy: listen tcp 0.0.0.0:5432: bind: address already in use
```

Different port indeed :+1:

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+   volumes:
+     - ./OpenOversight/:/usr/src/app/OpenOversight/
+   links:
+     - postgres:postgres
+   expose:
+     - "3000"
+   command: /usr/local/bin/gunicorn -w 4 -b 0.0.0.0:3000 app:app
+   ports:
+     - "3000:3000"
+     - "5432:5432"
```
Comment (r4v5) [inline docker-compose.yml]: As written, this binds the host's port 5432 to the `web` container's 5432, right? If I understand docker-compose syntax, to make the postgres server accessible to the host, this line would need to go up to a `ports` stanza on `postgres`, right? If we don't want to make it host-accessible, we can just remove this line. (You already updated `CONTRIB.md` to show executing psql within the container, so _theoretically_ we might just not want to expose it to the host at all by default?)

File: docker-compose.yml [hunk 2]
```diff
@@ -23,4 +23,4 @@ services:
    command: /usr/local/bin/gunicorn -w 4 -b 0.0.0.0:3000 app:app
    ports:
      - "3000:3000"
-     - "5432:5432"
+     - "5432:5433"
```
Comment (elimisteve) [inline docker-compose.yml]: @redshiftzero I think this should be "5433:5432" instead, so it's 5433 on the host.

**Comment by elimisteve** [comment]
After a couple runs I'm getting

```
[*] Populating database with test data...
/usr/local/lib/python2.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
[!] Encountered an unknown issue, exiting.
(psycopg2.IntegrityError) duplicate key value violates unique constraint "ix_departments_name"
DETAIL:  Key (name)=(Springfield Police Department) already exists.
 [SQL: 'INSERT INTO departments (name, short_name) VALUES (%(name)s, %(short_name)s) RETURNING departments.id'] [parameters: {'name': 'Springfield Police Department', 'short_name': 'SPD'}]
Makefile:10: recipe for target 'build' failed
make[1]: *** [build] Error 1
make[1]: Leaving directory '/home/user/django_projects/OpenOversight'
Makefile:6: recipe for target 'dev' failed
make: *** [dev] Error 2
```

**Comment by redshiftzero** [comment]
Ahhh test data script is not idempotent, whoops. Fixing

File: docker-compose.yml (hunk 2 above)
Comment (redshiftzero) [reply docker-compose.yml]: doh thanks

**Comment by redshiftzero** [comment]
This should be stampable since multiple peeps were using during the recent hackathon

**Comment by b-meson** [comment]
I didn't see the other commits on top of it so I will retest. assigned to myself.

**Comment by redshiftzero** [comment]
How did the testing go @b-meson?

**Comment by b-meson** [comment]
I wasn't able to run tests, looks like there is a hardcoded path `/vagrant/OpenOversight/` that we haven't fixed. Traceback below:

```
(env) zerohistory:OpenOversight bmeson$ make test
docker-compose run --rm web /usr/local/bin/pytest -v tests/
WARNING: The AWS_DEFAULT_REGION variable is not set. Defaulting to a blank string.
WARNING: The AWS_ACCESS_KEY_ID variable is not set. Defaulting to a blank string.
WARNING: The AWS_SECRET_ACCESS_KEY variable is not set. Defaulting to a blank string.
WARNING: The S3_BUCKET_NAME variable is not set. Defaulting to a blank string.
Starting openoversight_postgres_1 ... done
--------------------------------------------------------------------------------
INFO in __init__ [/usr/src/app/OpenOversight/app/__init__.py:57]:
OpenOversight startup
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/_pytest/config.py", line 337, in _importconftest
    mod = conftestpath.pyimport()
  File "/usr/local/lib/python2.7/site-packages/py/_path/local.py", line 686, in pyimport
    raise self.ImportMismatchError(modname, modfile, self)
ImportMismatchError: ('OpenOversight.tests.conftest', '/vagrant/OpenOversight/tests/conftest.py', local('/usr/src/app/OpenOversight/tests/conftest.py'))
ERROR: could not load /usr/src/app/OpenOversight/tests/conftest.py

make: *** [test] Error 4
```

**Comment by redshiftzero** [comment]
hey - pull the latest, blast away all `pyc` and `__pycache__` files and rebuild and this should be good to go

**Comment by b-meson** [comment]
`make test` now passes on this branch. Do I need to do anything else to QA this?

**Comment by redshiftzero** [comment]
woohoooooo :-) nope that's the ticket!

**Pull request merged**